From 085038ca5ac47d914cea322e25f12649cd1fe27b Mon Sep 17 00:00:00 2001
From: Yonggang Luo <luoyonggang@gmail.com>
Date: Fri, 3 Dec 2021 00:01:34 +0800
Subject: [PATCH] win32: Fixes building with msys2/mingw/clang32,clang64,ucrt64

Move `#define snprintf _snprintf` from nir/nir_lower_atomics_to_ssbo.c to nir.h

Fixes __MSVCRT_VERSION__ properly

Signed-off-by: Yonggang Luo <luoyonggang@gmail.com>
---
 meson.build                                  | 13 ++++++++++++-
 src/compiler/nir/nir.h                       |  2 ++
 src/compiler/nir/nir_lower_atomics_to_ssbo.c |  4 ----
 src/compiler/spirv/vtn_private.h             |  2 +-
 src/gallium/drivers/swr/swr_fence.cpp        |  2 +-
 5 files changed, 16 insertions(+), 7 deletions(-)

diff --git a/meson.build b/meson.build
index 8b8ffc95aab..ed7f93eac16 100644
--- a/meson.build
+++ b/meson.build
@@ -1089,7 +1089,18 @@ elif host_machine.system() == 'windows'
       '-DNOMINMAX',
     ]
   else
-    pre_args += ['-D__MSVCRT_VERSION__=0x0700']
+    # When the target is not mingw/ucrt
+    # NOTE: clang's stddef.h are conflict with mingw/ucrt's stddef.h
+    # So do not include headers that defined in clang for detecting
+    # _UCRT
+    if cc.compiles('''
+      #include <string.h>
+      #if defined(__MINGW32__) && defined(_UCRT)
+      #error
+      #endif
+      int main(void) { return 0; }''')
+      pre_args += ['-D__MSVCRT_VERSION__=0x0700']
+    endif
   endif
 elif host_machine.system() == 'openbsd'
   pre_args += '-D_ISOC11_SOURCE'
diff --git a/src/compiler/nir/nir.h b/src/compiler/nir/nir.h
index 019e60f9238..3a0f0e705c2 100644
--- a/src/compiler/nir/nir.h
+++ b/src/compiler/nir/nir.h
@@ -55,8 +55,10 @@
 #include "nir_opcodes.h"
 
 #if defined(_WIN32) && !defined(snprintf)
+#if (defined(_MSC_VER) && _MSC_VER < 1900) || (defined(__MINGW32__) && !defined(_UCRT))
 #define snprintf _snprintf
 #endif
+#endif
 
 #ifdef __cplusplus
 extern "C" {
diff --git a/src/compiler/nir/nir_lower_atomics_to_ssbo.c b/src/compiler/nir/nir_lower_atomics_to_ssbo.c
index c1799d0b9d0..448f63bdc7c 100644
--- a/src/compiler/nir/nir_lower_atomics_to_ssbo.c
+++ b/src/compiler/nir/nir_lower_atomics_to_ssbo.c
@@ -27,10 +27,6 @@
 #include "nir.h"
 #include "nir_builder.h"
 
-#if defined(_WIN32) && !defined(snprintf)
-#define snprintf _snprintf
-#endif
-
 /*
  * Remap atomic counters to SSBOs, starting from the shader's next SSBO slot
  * (info.num_ssbos).
diff --git a/src/compiler/spirv/vtn_private.h b/src/compiler/spirv/vtn_private.h
index 93985f54f86..24958cc13d7 100644
--- a/src/compiler/spirv/vtn_private.h
+++ b/src/compiler/spirv/vtn_private.h
@@ -41,7 +41,7 @@ struct vtn_builder;
 struct vtn_decoration;
 
 /* setjmp/longjmp is broken on MinGW: https://sourceforge.net/p/mingw-w64/bugs/406/ */
-#ifdef __MINGW32__
+#if defined(__MINGW32__) && !defined(_UCRT)
   #define vtn_setjmp __builtin_setjmp
   #define vtn_longjmp __builtin_longjmp
 #else
diff --git a/src/gallium/drivers/swr/swr_fence.cpp b/src/gallium/drivers/swr/swr_fence.cpp
index 4e2b2af874c..7803c77e425 100644
--- a/src/gallium/drivers/swr/swr_fence.cpp
+++ b/src/gallium/drivers/swr/swr_fence.cpp
@@ -29,7 +29,7 @@
 #include "swr_screen.h"
 #include "swr_fence.h"
 
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(__MINGW32__)
 #include <sched.h>
 #endif
 
-- 
2.31.1.windows.1

